keps/sig-apps/20190326-stateful-application-api.md

```diff
@@ -0,0 +1,790 @@
+  * Snapshots of a group of relevant persistent volumes used by the application.
+  * The configuration of the application and its sub-components, captured as a copy of the resource definitions of Kubernetes objects the application is comprised of.
+* **Application Backup**: a capture of the state of an application (both configuration and data) that is stored outside of a Kubernetes cluster. This may be an export of a local application snapshot to a remote storage.
+
+### API Structure
+
+The diagram below illustrates the overall Kubernetes API structure. Dotted arrows mean association through object names or object references. Solid arrows mean a binding between the two resource types involved. Green colored resources are the user-facing ones, blue colored ones are expected to be created and managed by system administrators, and light red colored ones are normally created as a result of performing the requested operations.
+
+![API structure](./k8s-stateful-app-mgmt-api.png)
+
+With this API structure, users can take a snapshot or backup of a `StatefulApplication`, via an `ApplicationSnapshot` or `ApplicationBackup`, respectively. In the reverse direction, both `ApplicationSnapshot`s and `ApplicationBackup`s can be used as data sources in application restoration or clone. An `ApplicationSnapshot` can be exported outside of a cluster as an `ApplicationBackup`, and an `ApplicationBackup` can be imported into the cluster as an `ApplicationSnapshot`. This allows users to use all the stateful application data management semantics in a Kubernetes idiomatic manner. 
+
+Although we introduced the concept of application backups and mentioned the `ApplicationBackup` CRD above, this KEP focuses on details of CRDs related to application snapshot and restoration/clone operations and introduces the following CRDs:
+* `StatefulApplication` and `StatefulApplicationClass` for representation and description of stateful applications and for application restoration/clone support,
+* and `ApplicationSnapshot`, `ApplicationSnapshotClass`, and `ApplicationSnapshotContent` for application snapshot support,
+* and `ApplicationRollback` for application rollback support.
+
+Details of CRDs related to application backups including `ApplicationBackup`, `ApplicationBackupClass`, and `ApplicationBackupContent` will be introduced in a subsequent KEP or an update to this KEP.
+
+### Stateful Application
+
+The API is centered around the notion of stateful applications modeled using a CRD named `StatefulApplication`. The `StatefulApplication` CRD leverages the [Application CRD](https://github.com/kubernetes-sigs/application) for  description and aggregation of components and status of an application. Additionally, it supports data management semantics such as snapshot, backup, restoration, and clone for the application it models. 
```

Comment:
Is there a plan for managing the dependency between this work and the Application CRD, e.g. across versions of that CRD? It might be difficult to deal with changes in the App CRD spec, and users will have to install that CRD separately to the ones described here in their cluster. The same concerns probably go for the ExecutionHook CRD.

Reply:
That's a good question. We don't have one currently, but I agree that we need a proposal for dependency management between related CRDs.